### ISSUE TITLE
EnterCriticalSection not hookable on Windows

### ISSUE THREAD

**Issue by sebastianwalla** [angr/angr#847]
I have a binary which was compiled using Microsofts Visual Studio 2017 C++ compiler. When trying to symbolically execute this binary on a Windows 10 32Bit System it has problems with the printf function
When trying to hook the EnterCriticalSection function in angr it seems like still the real library function is called instead of the sim-procedure. Furthermore it seems like I can only hook either EnterCriticalSection or LeaveCriticalSection, since they both are mapped to the same key in the project._sim_procedures dictionary

I tried to produce a minimal example to reproduce this issue:
Source code of demo binary:
```
#include <stdlib.h>
#include <stdio.h>


int __cdecl main(int argc, char **argv)
{
    printf("Hello World!");
}
```

Source-code of angr script:
```
import angr
import json
import logging
import time


#logging.getLogger('angr').setLevel('DEBUG')

DEBUG=True

def custom_log(msg):
    if DEBUG:
        print msg

# currently we lock nothing
#https://msdn.microsoft.com/de-de/library/windows/desktop/ms682608(v=vs.85).aspx
class EnterCriticalSection(angr.SimProcedure):
    def run(self,lpCriticalSection):
        self.cc.CALLEE_CLEANUP = True
        custom_log("EnterCriticalSection called")
        return # returns nothing


# currently we lock nothing
#https://msdn.microsoft.com/de-de/library/windows/desktop/ms682608(v=vs.85).aspx
class LeaveCriticalSection(angr.SimProcedure):
    def run(self,lpCriticalSection):
        self.cc.CALLEE_CLEANUP = True
        custom_log("LeaveCriticalSection called")
        return # returns nothing


def hook_functions(proj):
    # hook into interesting library functions here
    proj.hook_symbol("EnterCriticalSection", EnterCriticalSection())
    proj.hook_symbol("LeaveCriticalSection", LeaveCriticalSection()) # if this is executed the EnterCriticalSection hook is replaced
  

def analyze_program():
    path_to_binary = "./main.exe"  # todo make this use args

    custom_entry_point = 0x401000  # start of main


    proj = angr.Project(path_to_binary) # for testing on win32 system
    
    hook_functions(proj)

 
    state = proj.factory.call_state(custom_entry_point)  
    simgr = proj.factory.simgr(state)  
    step_until_branch(simgr)
    return

analyze_program()
```

**Comment by schieb** [comment]
I can only build up to v141. For v141, the compiler does not emit calls to `EnterCriticalSection`. Could you please upload the binary?

**Comment by schieb** [comment]
> Furthermore it seems like I can only hook either EnterCriticalSection or LeaveCriticalSection, since they both are mapped to the same key in the project._sim_procedures dictionary

I don't understand what you mean. Dictionary keys are unique, therefore two values (SimProcedure instances) cannot be mapped to the same key (address). It could be the case that both symbols refer to the same object but that is unlikely. Could you clarify what you mean?

**Comment by sebastianwalla** [comment]
Actually I can not reproduce the issue with the hooking to the same address with this source code, but with another one: [Binary where only hooking EnterCriticalSection or LeaveCriticalSection is possible (called main.zip)](https://github.com/angr/angr/files/1691689/main.zip)

> I don't understand what you mean. Dictionary keys are unique, therefore two values (SimProcedure instances) cannot be mapped to the same key (address). It could be the case that both symbols refer to the same object but that is unlikely. Could you clarify what you mean?

In my `hook_functions` function I first try to hook the EnterCriticalSection symbol using `hook_symbol` like described [here](https://docs.angr.io/docs/simprocedures.html) in the documentation.
I debugged this and noticed that this will map in `proj._sim_procedures` from `83890312` to the SimProcedure EnterCriticalSection. If I then execute the next line it will try to hook the LeaveCriticalSection symbol, but now the entry in the map `proj._sim_procedures` for the key `83890312` will point to the LeaveCriticalSection SimProcedure and not the EnterCriticalSection Procedure anymore. So if one tries to hook both Symbols at a time with this binary it will not work.

Regarding the problem that if only the EnterCriticalSection symbol is hooked and not the LeaveCriticalSection symbol, but still the library function instead of the SimProcedure is called:
This is the corresponding binary to the above posted source_code 
[printf_demo.zip](https://github.com/angr/angr/files/1691719/printf_demo.zip)

I debugged the symbolic execution of this binary and think it is because somehow the VEX engine will not stop at the instruction `call EnterCriticalSection` but instead go into the actual library and execute it. If I use as a load_options `{'auto_load_libs':False}` for the `Project-Constructor` the EnterCriticalSection SimProcedure is called just as expected and not the library function.
